aten/src/ATen/native/cuda/IndexKernel.cu

```diff
@@ -314,50 +268,91 @@ Tensor & masked_select_out_cuda(const Tensor & self, const Tensor & mask, Tensor
   return masked_select_out_cuda_impl(result, self, mask);
 }
 
-void take_out_cuda_template(Tensor& output, const Tensor& input, const Tensor& index) {
-  TORCH_CHECK(output.device().type() == at::kCUDA, "device type of output (", output.device().type(), ") is not GPU");
```

Comment:
Should I add any other of these asserts to the  `take_out` / `put_` functions? @mruberry

Reply:
You should add TORCH_CHECK to make sure devices match, it's not necessary to check that all devices are cuda or that layout is strided (dispatcher will take care of that). 
Do you know why output tests from OpInfo didn't fail? They are supposed to check that. Did the job just not get to them?

Reply:
That's already been checked here:
https://github.com/pytorch/pytorch/blob/dc8ce32809122407af25cb8d910936be1cda0fcf/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L404-L407
Same for `put_`.

So, if this is alright, I think that this PR should be all done now :) Let's try to merge before there are any other merge conflicts, as this PR touches almost 30 files.

Reply:
That's interesting. How was this not caught here? It's important that test work properly for exactly this reason.

https://github.com/pytorch/pytorch/blob/a4125876c9b62cd59156e5a7e2bcd6c0b6d42897/test/test_ops.py#L593

Reply:
I have no idea. I will investigate it in a bit. But this point is not blocking for this PR.

Reply:
OK. @ngimel and I should be able to take a look at this later today, but we'd really appreciate a post-mortem on how that test failed.

Reply:
Nothing to worry about the tests. It turns out that `TensorIterator` was already checking that, as it was throwing in that case a:
`Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!`